src/Geography.cpp

```diff
@@ -0,0 +1,236 @@
+        buffer.putDouble(x);
+        buffer.putDouble(y);
+        buffer.putDouble(z);
+    }
+    char zeros[38];
+    memset(zeros, 0, sizeof(zeros));
+    buffer.put(zeros, sizeof(zeros));
+}
+
+bool Geography::approximatelyEqual(const Geography &rhs, double epsilon) const {
+    if (numRings() != rhs.numRings()) {
+        return false;
+    }
+    for (int idx = 0; idx < m_rings.size(); idx += 1) {
+        if (!getRing(idx).approximatelyEqual(rhs.getRing(idx), epsilon)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+int32_t Geography::getSerializedSize() const {
+    // Null Geography values are
+    // just 4 bytes long - just enough
+    // for the -1.
+    if (isNull()) {
+        return 4;
+    }
+    int32_t answer
+        = 4 // size
+        + 3 // 3 internal bytes
+        + 4 // Count of the number of rings
+        + 33; // Trailing overhead
+    for (RingConstIterator idx = m_rings.begin();
+         idx != m_rings.end();
+         idx++) {
+        // Each ring has 3 coordinates for each point
+        // plus 5 + 38 bytes of overhead.  Note that we
+        // will not serialize the duplicated first point.
+        answer += (idx->numPoints()-1)*3*sizeof(double) + 5 + 38;
+    }
+    return answer;
+}
+
+int32_t Geography::serializeTo(ByteBuffer &buffer) const {
+    // Remember where we start, so that we can
+    // write the size.  We keep track of how much
+    // we actually write and return it, so that the
+    // caller can check that this and getSerializedSize
+    // agree.
+    int32_t start = buffer.position();
+    if (isNull()) {
+        buffer.putInt32(-1);
+        return 4;
+    }
+
+    // Save space for the size.  We don't
+    // really know what it is now, so we
+    // well seek back to it after we are done.
+    buffer.putInt32(0);
+    // We write three bytes of overhead, followed by the
+    // number of rings.
+    buffer.putInt8(0);
+    buffer.putInt8(1);
+    buffer.putInt8(0);
+    buffer.putInt32(m_rings.size());
+    bool reverseit = false;
+    for (RingConstIterator idx = m_rings.begin();
+         idx != m_rings.end();
+         idx++) {
+        // Serialize each ring.
+        idx->serializeTo(buffer, reverseit);
+        reverseit = true;
+    }
+
+    // We write 33 bytes of overhead.
+    char zeros[33];
+    memset(zeros, 0, sizeof(zeros));
+    buffer.put(zeros, sizeof(zeros));
+
+    // Write the size where we started.
+    int32_t size = buffer.position() - start;
+    buffer.putInt32(start, size-4);
+    return size;
+}
+
+int32_t Geography::deserializeFrom(ByteBuffer &aData,
+                                   int32_t first_offset,
+                                   bool &aWasNull)
+{
+    int32_t offset = first_offset;
+    // The first four bytes are the length.  We read them
+    // only so we can validate the read.  The length does
+    // not include them.
+    int32_t len = aData.getInt32(offset);
+    if (len == -1) {
+        aWasNull = true;
```

Comment:
I don't understand this?  This routine knows whether the thing its reading is null
or not.  It needs to return that to the caller.  The boolean which aWasNull is a
reference to is the value isNull returns.  So it has to be set somewhere.